kvbc/src/categorization/block_merkle_category.cpp

```diff
@@ -0,0 +1,398 @@
+  auto hashed_keys = hashedKeys(keys);
+  std::vector<std::optional<BlockId>> versions;
+  versions.reserve(keys.size());
+  multiGetLatestVersion(hashed_keys, versions);
+
+  // Generate the set of versioned keys for all keys that have latest versions
+  auto versioned_keys = std::vector<Buffer>{};
+  auto found_versions = std::vector<BlockId>{};
+  for (auto i = 0u; i < hashed_keys.size(); i++) {
+    if (versions[i]) {
+      found_versions.push_back(*versions[i]);
+      versioned_keys.push_back(serialize(VersionedKey{KeyHash{hashed_keys[i]}, *versions[i]}));
+    }
+  }
+
+  values.clear();
+  // Optimize for all keys existing (having latest versions)
+  if (versioned_keys.size() == hashed_keys.size()) {
+    return multiGet(versioned_keys, found_versions, values);
+  }
+
+  // Retrieve only the keys that have latest versions
+  auto retrieved_values = std::vector<std::optional<MerkleValue>>{};
+  retrieved_values.reserve(versioned_keys.size());
+  multiGet(versioned_keys, found_versions, retrieved_values);
+
+  // Merge any keys that didn't have latest versions along with the retrieved keys.
+  auto value_index = 0u;
+  for (auto& version : versions) {
+    if (version) {
+      values.push_back(retrieved_values[value_index]);
+      ++value_index;
+    } else {
+      values.push_back(std::nullopt);
+    }
+  }
+  ConcordAssertEQ(values.size(), keys.size());
+}
+
+void BlockMerkleCategory::putKeys(NativeWriteBatch& batch,
+                                  uint64_t block_id,
+                                  const std::vector<KeyHash>& hashed_added_keys,
+                                  const std::vector<KeyHash>& hashed_deleted_keys,
+                                  BlockMerkleInput& updates) {
+  auto kv_it = updates.kv.begin();
+  for (auto key_it = hashed_added_keys.begin(); key_it != hashed_added_keys.end(); key_it++) {
+    // Write the versioned key/value pair used for direct key lookup
+    batch.put(BLOCK_MERKLE_KEYS_CF, serialize(VersionedKey{*key_it, block_id}), kv_it->second);
+
+    // Put the latest version of the key
+    batch.put(BLOCK_MERKLE_LATEST_KEY_VERSION, key_it->value, serialize(LatestKeyVersion{block_id}));
+
+    kv_it++;
+  }
+
+  for (auto key_it = hashed_deleted_keys.begin(); key_it != hashed_deleted_keys.end(); key_it++) {
+    // Add the new tombstone to the key versions. We use 0 here to avoid the serialization overhead of a
+    // variant/optional.
+    batch.put(BLOCK_MERKLE_LATEST_KEY_VERSION, key_it->value, serialize(LatestKeyVersion{0}));
+  }
+}
+
+void BlockMerkleCategory::putMerkleNodes(NativeWriteBatch& batch,
+                                         sparse_merkle::UpdateBatch&& update_batch,
+                                         uint64_t tree_version) {
+  for (const auto& [leaf_key, leaf_val] : update_batch.leaf_nodes) {
+    auto ser_key = serialize(leafKeyToVersionedKey(leaf_key));
+    batch.put(BLOCK_MERKLE_LEAF_NODES_CF, ser_key, leaf_val.value.string_view());
+  }
```

Comment:
If I understand correctly we should get only one leaf in ```update_batch.leaf_nodes``` if so does it worth to assert?

Reply:
Sure. It's at least good documentation at a minimum.